### ISSUE TITLE
6.0.4/dtest-debug: test_remove_node_during_local_index_build - coredump - 

### ISSUE THREAD

**Issue by mykaul** [scylladb/scylladb#20787]
Link - https://jenkins.scylladb.com/job/scylla-6.0/job/dtest-debug/26/testReport/secondary_indexes_test/TestLocalIndexes/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split033___test_remove_node_during_local_index_build/ 
Log snippet:
```
INFO  2024-09-22 07:27:01,088 [shard 0:strm] stream_session - [Stream #0ee12fe0-78b4-11ef-ab7b-ed5f9e90bc69] Executing streaming plan for Tablet migration-ks-index-0 with peers=127.0.63.5, slave
INFO  2024-09-22 07:27:01,090 [shard 0:strm] stream_session - [Stream #0ee12fe0-78b4-11ef-ab7b-ed5f9e90bc69] Skip sending ks=ks, cf=cf, reader contains no data, with new rpc streaming
INFO  2024-09-22 07:27:01,123 [shard 1:strm] stream_session - [Stream #0ee12fe0-78b4-11ef-ab7b-ed5f9e90bc69] Start sending ks=ks, cf=cf, estimated_partitions=49664, with new rpc streaming
INFO  2024-09-22 07:27:01,374 [shard 0:strm] stream_session - [Stream #0f0cfad0-78b4-11ef-ab7b-ed5f9e90bc69] Executing streaming plan for Tablet migration-ks-index-0 with peers=127.0.63.5, slave
INFO  2024-09-22 07:27:01,377 [shard 1:strm] stream_session - [Stream #0f0cfad0-78b4-11ef-ab7b-ed5f9e90bc69] Skip sending ks=ks, cf=cf, reader contains no data, with new rpc streaming
INFO  2024-09-22 07:27:01,385 [shard 0:strm] stream_session - [Stream #0f0cfad0-78b4-11ef-ab7b-ed5f9e90bc69] Start sending ks=ks, cf=cf, estimated_partitions=25088, with new rpc streaming
INFO  2024-09-22 07:27:14,444 [shard 0:strm] stream_session - [Stream #0f0cfad0-78b4-11ef-ab7b-ed5f9e90bc69] Streaming plan for Tablet migration-ks-index-0 succeeded, peers={127.0.63.5}, tx=5071 KiB, 388.00 KiB/s, rx=0 KiB, 0.00 KiB/s
INFO  2024-09-22 07:27:17,235 [shard 0:strm] table - Cleaned up tablet 2 of table ks.cf successfully.
INFO  2024-09-22 07:27:32,976 [shard 0:strm] stream_session - [Stream #0ee12fe0-78b4-11ef-ab7b-ed5f9e90bc69] Streaming plan for Tablet migration-ks-index-0 succeeded, peers={127.0.63.5}, tx=5034 KiB, 157.88 KiB/s, rx=0 KiB, 0.00 KiB/s
INFO  2024-09-22 07:27:33,761 [shard 0:comp] compaction - [Compact system.raft 225ad800-78b4-11ef-b09a-76ded47f9048] Compacting [/jenkins/workspace/scylla-6.0/dtest-debug/scylla/.dtest/dtest-9uytahmo/test/node4/data/system/raft-3e17774c57f539939625327cbafbf5bb/me-3gjr_0kpx_413rk2oolhtfvqa2bs-big-Data.db:level=0:origin=memtable,/jenkins/workspace/scylla-6.0/dtest-debug/scylla/.dtest/dtest-9uytahmo/test/node4/data/system/raft-3e17774c57f539939625327cbafbf5bb/me-3gjr_0kmr_3rgjk2oolhtfvqa2bs-big-Data.db:level=0:origin=compaction]
INFO  2024-09-22 07:27:33,930 [shard 0:comp] compaction - [Compact system.raft 225ad800-78b4-11ef-b09a-76ded47f9048] Compacted 2 sstables to [/jenkins/workspace/scylla-6.0/dtest-debug/scylla/.dtest/dtest-9uytahmo/test/node4/data/system/raft-3e17774c57f539939625327cbafbf5bb/me-3gjr_0kpx_4jr282oolhtfvqa2bs-big-Data.db:level=0]. 541kB to 534kB (~98% of original) in 129ms = 4MB/s. ~256 total partitions merged to 1.
INFO  2024-09-22 07:27:34,404 [shard 1:strm] table - Cleaned up tablet 3 of table ks.cf successfully.
terminate called after throwing an instance of 'std::out_of_range'
  what():  Storage wasn't found for tablet 3 of table ks.cf
Aborting on shard 1.
Backtrace:
  0xd9915ca
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x488563c
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x4885293
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x46bffc8
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x46f0b95
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x47eb7a9
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x47ebaee
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x47eb8fe
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libc.so.6+0x3dbaf
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libc.so.6+0x8e883
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libc.so.6+0x3dafd
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libc.so.6+0x2687e
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libstdc++.so.6+0xa4d38
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libstdc++.so.6+0xb4f6b
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libstdc++.so.6+0xb4fd6
  0xdb2b0ea
  0xed8365f
  0xebfb690
  0xec7d4d0
  0xecdb039
  0xed8c39f
  0xed662f7
  0xed5468d
  0xee256f6
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x46e2afe
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x46eb020
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x46ef042
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x47ee4b5
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x47ec650
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x47ec540
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x47ebcac
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libseastar.so+0x44913f8
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libc.so.6+0x8c946
  /jenkins/workspace/scylla-6.0/dtest-debug/scylla/.ccm/scylla-repository/7d421abec43db7dacdf135ee27ce1eb0eb5cd570/libreloc/libc.so.6+0x11296f
```

decoded:
```
Backtrace:
[Backtrace #0]
__interceptor_backtrace at ??:?
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/debug/seastar/./seastar/include/seastar/util/backtrace.hh:68
seastar::backtrace_buffer::append_backtrace() at ./build/debug/seastar/./seastar/src/core/reactor.cc:825
seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/debug/seastar/./seastar/src/core/reactor.cc:855
seastar::print_with_backtrace(char const*, bool) at ./build/debug/seastar/./seastar/src/core/reactor.cc:867
seastar::sigabrt_action() at ./build/debug/seastar/./seastar/src/core/reactor.cc:4071
seastar::install_oneshot_signal_handler<6, (void (*)())(&seastar::sigabrt_action)>()::{lambda(int, siginfo_t*, void*)#1}::operator()(int, siginfo_t*, void*) const at ./build/debug/seastar/./seastar/src/core/reactor.cc:4047
seastar::install_oneshot_signal_handler<6, (void (*)())(&seastar::sigabrt_action)>()::{lambda(int, siginfo_t*, void*)#1}::__invoke(int, siginfo_t*, void*) at ./build/debug/seastar/./seastar/src/core/reactor.cc:4043
/data/scylla-s3-reloc.cache/by-build-id/c2e9544623fbc78467daf8dfadb6e83be43176dd/extracted/scylla/libreloc/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=9148cab1b932d44ef70e306e9c02ee38d06cad51, for GNU/Linux 3.2.0, not stripped

__GI___sigaction at :?
__pthread_kill_implementation at ??:?
__GI_raise at :?
__GI_abort at :?
/data/scylla-s3-reloc.cache/by-build-id/c2e9544623fbc78467daf8dfadb6e83be43176dd/extracted/scylla/libreloc/libstdc++.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, BuildID[sha1]=f47a734d459c3340632cd52c18936fbde51abba4, stripped

__cxa_throw_bad_array_new_length at ??:?
std::rethrow_exception(std::__exception_ptr::exception_ptr) at ??:?
std::terminate() at ??:?
__clang_call_terminate at main.cc:?
replica::tablet_storage_group_manager::storage_group_for_token(dht::token) const at ./replica/table.cc:773
replica::table::storage_group_for_token(dht::token) const at ./replica/table.cc:987
 (inlined by) replica::table::add_memtables_to_reader_list(std::vector<flat_mutation_reader_v2, std::allocator<flat_mutation_reader_v2> >&, seastar::lw_shared_ptr<schema const> const&, reader_permit const&, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr const&, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>, std::function<void (unsigned long)>) const at ./replica/table.cc:194
replica::table::make_reader_v2_excluding_staging(seastar::lw_shared_ptr<schema const>, reader_permit, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>) const at ./replica/table.cc:3277
replica::table::as_mutation_source_excluding_staging() const::$_0::operator()(seastar::lw_shared_ptr<schema const>, reader_permit, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>) const at ./replica/table.cc:3431
 (inlined by) flat_mutation_reader_v2 std::__invoke_impl<flat_mutation_reader_v2, replica::table::as_mutation_source_excluding_staging() const::$_0&, seastar::lw_shared_ptr<schema const>, reader_permit, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag> >(std::__invoke_other, replica::table::as_mutation_source_excluding_staging() const::$_0&, seastar::lw_shared_ptr<schema const>&&, reader_permit&&, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr&&, seastar::bool_class<streamed_mutation::forwarding_tag>&&, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>&&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:61
 (inlined by) std::enable_if<is_invocable_r_v<flat_mutation_reader_v2, replica::table::as_mutation_source_excluding_staging() const::$_0&, seastar::lw_shared_ptr<schema const>, reader_permit, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag> >, flat_mutation_reader_v2>::type std::__invoke_r<flat_mutation_reader_v2, replica::table::as_mutation_source_excluding_staging() const::$_0&, seastar::lw_shared_ptr<schema const>, reader_permit, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag> >(replica::table::as_mutation_source_excluding_staging() const::$_0&, seastar::lw_shared_ptr<schema const>&&, reader_permit&&, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr&&, seastar::bool_class<streamed_mutation::forwarding_tag>&&, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>&&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:114
 (inlined by) std::_Function_handler<flat_mutation_reader_v2 (seastar::lw_shared_ptr<schema const>, reader_permit, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>), replica::table::as_mutation_source_excluding_staging() const::$_0>::_M_invoke(std::_Any_data const&, seastar::lw_shared_ptr<schema const>&&, reader_permit&&, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr&&, seastar::bool_class<streamed_mutation::forwarding_tag>&&, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>&&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:290
std::function<flat_mutation_reader_v2 (seastar::lw_shared_ptr<schema const>, reader_permit, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>)>::operator()(seastar::lw_shared_ptr<schema const>, reader_permit, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>) const at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:591
mutation_source::make_reader_v2(seastar::lw_shared_ptr<schema const>, reader_permit, interval<dht::ring_position> const&, query::partition_slice const&, tracing::trace_state_ptr, seastar::bool_class<streamed_mutation::forwarding_tag>, seastar::bool_class<mutation_reader::partition_range_forwarding_tag>) const at ././readers/mutation_source.hh:142
replica::table::do_push_view_replica_updates(seastar::shared_ptr<db::view::view_update_generator>, seastar::lw_shared_ptr<schema const>, mutation, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000000000l> > >, mutation_source, tracing::trace_state_ptr, reader_concurrency_semaphore&, enum_set<super_enum<query::partition_slice::option, (query::partition_slice::option)0, (query::partition_slice::option)1, (query::partition_slice::option)2, (query::partition_slice::option)3, (query::partition_slice::option)4, (query::partition_slice::option)5, (query::partition_slice::option)6, (query::partition_slice::option)7, (query::partition_slice::option)8, (query::partition_slice::option)9, (query::partition_slice::option)10, (query::partition_slice::option)11, (query::partition_slice::option)12, (query::partition_slice::option)13> >) const at ./replica/table.cc:3392
std::__n4861::coroutine_handle<seastar::internal::coroutine_traits_base<row_locker::lock_holder>::promise_type>::resume() const at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/coroutine:240
 (inlined by) seastar::internal::coroutine_traits_base<row_locker::lock_holder>::promise_type::run_and_dispose() at ././seastar/include/seastar/core/coroutine.hh:83
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at ./build/debug/seastar/./seastar/src/core/reactor.cc:2690
seastar::reactor::run_some_tasks() at ./build/debug/seastar/./seastar/src/core/reactor.cc:3152
seastar::reactor::do_run() at ./build/debug/seastar/./seastar/src/core/reactor.cc:3320
seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0::operator()() const at ./build/debug/seastar/./seastar/src/core/reactor.cc:4563
void std::__invoke_impl<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&>(std::__invoke_other, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:61
std::enable_if<is_invocable_r_v<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&>, void>::type std::__invoke_r<void, seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&>(seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/invoke.h:111
std::_Function_handler<void (), seastar::smp::configure(seastar::smp_options const&, seastar::reactor_options const&)::$_0>::_M_invoke(std::_Any_data const&) at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/std_function.h:290
seastar::posix_thread::start_routine(void*) at ./build/debug/seastar/./seastar/src/core/posix.cc:90
start_thread at ??:?
__clone3 at :?
```

Could be https://github.com/scylladb/scylladb/pull/20628 that was not yet backported to 6.0?

**Comment by mykaul** [comment]
@Deexie - please take a look.

**Comment by Deexie** [comment]
This isn't related to https://github.com/scylladb/scylladb/pull/20628. It looks like a data race of tablet cleanup and some caller of `database::apply`, but I don't know which one yet

**Comment by raphaelsc** [comment]
this seems to happen in the context of view update from staging. I think it's fixed by 08b29460fc9578ce3d998d9dd96ea260f1c8dc5f, which 6.0.4 misses. not backported to 6.0 yet, only present in 6.1 branch. /cc @mlitvk

**Comment by mykaul** [comment]
> this seems to happen in the context of view update from staging. I think it's fixed by [08b2946](https://github.com/scylladb/scylladb/commit/08b29460fc9578ce3d998d9dd96ea260f1c8dc5f), which 6.0.4 misses. not backported to 6.0 yet, only present in 6.1 branch. /cc [@mlitvk](https://github.com/mlitvk)

@piodul 0 is the above on track to 6.0?

**Comment by piodul** [comment]
> @piodul 0 is the above on track to 6.0?

We didn't plan to backport that PR. I will do the backport now.

**Comment by piodul** [comment]
The backport landed on `branch-6.0` - closing.
